feat: add managed EFS to copilot

bvtujo opened the pull request:
<!-- Provide summary of changes -->
This PR adds the ability to specify managed EFS filesystems at the environment level by defining a magic volume in a service manifest. 

For example:

```yaml
name: frontend
type: Load Balanced Web Service

...

storage:
  volumes:
    magicEFSVolume:
      path: "/etc/efs"
      read_only: false
      efs: copilot # "managed" also works.
```

Or, you can configure the UID and GID of the access point manually:
```yaml
storage:
  volumes:
    magicEFSVolume:
      path: "/etc/efs"
      read_only: false
      efs:
        id: managed
        uid: 10000
        gid: 100000
```

This triggers the creation of a shared volume at the environment level, where each service or job has its own directory and its own access point. Access is gated via IAM policies. 

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bvtujo committed:
chore: add logic to only update env if requires EFS

bvtujo committed:
chore: add custom unmarshaler to EFS

bvtujo committed:
chore: enable managed EFS

bvtujo committed:
chore: validate/autogenerate UID/GID

bvtujo commented:
Integ tests are forthcoming later today.

bvtujo committed:
chore: fix staticcheck error

bvtujo committed:
chore: add tests and fix failing integ tests

bvtujo committed:
chore: update templates to use EnvController outputs

bvtujo committed:
chore: fix cloudformation errors

thstart commented:
I followed the instructions here:
https://aws.github.io/copilot-cli/docs/developing/storage/

```
storage:
  volumes:
    copilotVolume: # This is a variable key and can be set to arbitrary strings.
      path: '/app/db'
      read_only: false # Set to false if your service needs write access. 
      efs:
        id: fs-123456
```

Then:
```
$ aws cloudformation deploy 
    --stack-name efs-cfn \
    --template-file ecs.yml
    --parameter-overrides App=${YOUR_APP} Env=${YOUR_ENV}
```

Got the EFS filesystem ID
Then ran copilot svc deploy

Inside the Dockerfile I have this instructions:
COPY . ./

RUN cp -r /app/efs /app/db/

I expect the folder app/efs to be copied to the EFS /app/db/
but it is not. 

What is the solution?

bvtujo committed:
chore: add metadata to EFS filesystem and access point

iamhopaul123 commented:
Hello @thstart. Can you try to do the copy at docker run time instead of build time? Like having `CMD cp -r /app/efs /app/db/`?

thstart commented:
> Hello @thstart. Can you try to do the copy at docker run time instead of build time? Like having `CMD cp -r /app/efs /app/db/`?

Like this:
CMD cp -r /app/efs /app/db/
and immediately after that:
CMD ["/app/server"]

Correct?

Here is what I'm doing:
=========
Dockerfile:
=========
```
FROM golang:1.16-buster as build-env

WORKDIR /app

COPY . ./

RUN CGO_ENABLED=0 GOARCH=amd64 GOOS=linux \
    go build -a -installsuffix cgo \
    server.go

FROM gcr.io/distroless/base
COPY --from=build-env /app/server /app/server
COPY --from=build-env /app/efs/ /app/efs/

COPY --from=build-env /app/efs/STATE/ /app/db/STATE/
COPY --from=build-env /app/efs/COUNTY/ /app/db/COUNTY/

WORKDIR /app

EXPOSE 80
# Run the web service on container startup.
CMD ["/app/server"]


```
I expect /app/db/ to be the mounted EFS.

bvtujo commented:
> > Hello @thstart. Can you try to do the copy at docker run time instead of build time? Like having `CMD cp -r /app/efs /app/db/`?
> 
> 
> 
> Like this:
> 
> CMD cp -r /app/efs /app/db/
> 
> and immediately after that:
> 
> CMD ["/app/server"]
> 
> 
> 
> Correct?

You can only have one `CMD` in a Dockerfile so you may want to wrap your server in a simple script: 

```sh
# run.sh
#!/bin/sh

cp -r /app/efs-staging /app/efs

/app/server
```

```Dockerfile
# Dockerfile

...
CMD sh run.sh

```

thstart commented:
> > > Hello @thstart. Can you try to do the copy at docker run time instead of build time? Like having `CMD cp -r /app/efs /app/db/`?
> > 
> > 
> > Like this:
> > CMD cp -r /app/efs /app/db/
> > and immediately after that:
> > CMD ["/app/server"]
> > Correct?
> 
> You can only have one `CMD` in a Dockerfile so you may want to wrap your server in a simple script:
> 
> ```shell
> # run.sh
> #!/bin/sh
> 
> cp -r /app/efs-staging /app/efs
> 
> /app/server
> ```
> 
> ```dockerfile
> # Dockerfile
> 
> ...
> CMD sh run.sh
> ```

I do something like that:
COPY --from=build-env /app/server /app/server
because the folder is in the "build-env" and I want the final image to not 
have this folder. 

This command assumes the folder is inside the container:
cp -r /app/efs-staging /app/efs

The reason is I'm trying to find easier solution for transferring the 
folders to the EFS. With S3 it is much easier. If that solution is working
I suppose can use a separate docker file with the db folders just to 
upload them to the EFS and the other container can run the service 
using the EFS.

bvtujo committed:
test: add integ tests for managed EFS

thstart commented:
> > > Hello @thstart. Can you try to do the copy at docker run time instead of build time? Like having `CMD cp -r /app/efs /app/db/`?
> > 
> > 
> > Like this:
> > CMD cp -r /app/efs /app/db/
> > and immediately after that:
> > CMD ["/app/server"]
> > Correct?
> 
> You can only have one `CMD` in a Dockerfile so you may want to wrap your server in a simple script:
> 
> ```shell
> # run.sh
> #!/bin/sh
> 
> cp -r /app/efs-staging /app/efs
> 
> /app/server
> ```
> 
> ```dockerfile
> # Dockerfile
> 
> ...
> CMD sh run.sh
> ```

I'm using Copilot because it creates all permissions and setting and is working very reliably.

I would like to publish my container using the EFS at AWS Marketplace.
How to use Copilot to make that task easier? For example mounting the EFS to the container?

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -40,6 +41,10 @@ var (
 	errNoContainerPath = errors.New(`"path" cannot be empty`)
 	errNoSourceVolume  = errors.New(`"source_volume" cannot be empty`)
+
+	errUIDWithNonManagedFS = errors.New(`UID and GID cannot be specified with non-managed EFS`)
```
```suggestion
	errUIDWithNonManagedFS = errors.New("UID and GID cannot be specified with non-managed EFS")
```

bvtujo committed:
Merge branch 'mainline' into managed-efs

bvtujo committed:
chore: add managed efs to scheduled jobs

Lou1415926 commented on internal/pkg/manifest/storage.go:
```diff
@@ -28,13 +38,87 @@ type SidecarMountPoint struct {
+// EFSConfigOrID contains custom unmarshaling logic for the `efs` field in the manifest.
+type EFSConfigOrID struct {
+	Config EFSVolumeConfiguration
+	ID     string
```
Maybe we can add a comment here, especially for the `ID` field to note that it represents the same info as `FileSystemID`? Or what do you think of having the same name for both the ID variable and the Config.FileSystemID variable?

bvtujo replied:
This is actually going to be refactored out in my next PR 🤪 in favor of 

```go
type EFSConfigOrBool struct {
    Config  EfsVolumeConfiguration
    Enabled *bool
}
```

which I think will render this confusion moot.

Lou1415926 commented on internal/pkg/manifest/storage.go:
```diff
@@ -28,13 +38,87 @@ type SidecarMountPoint struct {
+		return aws.String(fromID), nil
+	}
+	fromConfig := aws.StringValue(e.Config.FileSystemID)
+	if fromID != "" && fromID != fromConfig {
```
Is it possible that both `fromID` and `fromConfig` are empty string? Should we return an error since id is required?

bvtujo replied:
I believe error checking is handled elsewhere; it's fine if this returns `nil, nil`. We want to support the possibility of having empty volumes, so this ensures that it's possible to have `FSID()` gracefully return if no ID is set.

Lou1415926 commented on internal/pkg/manifest/storage_test.go:
```diff
@@ -0,0 +1,113 @@
+	EFS *EFSConfigOrID `yaml:"efs"`
+}
+
+func Test_UnmarshalEFS(t *testing.T) {
```
Should we include a few more tests that cover error cases here? For example, in the case of 
```
efs: copilot
   id: managed
```
or 
```
efs: copilot
   id: random_id
```
etc.

bvtujo replied:
These edge cases will go away in the next refactor PR; I vote to leave this as is if that's okay.

Lou1415926 commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -251,41 +261,117 @@ func convertEFSPermissions(input map[string]manifest.Volume) ([]*template.EFSPer
+func convertManagedFSInfo(wl manifest.Workload, input map[string]manifest.Volume) (*template.ManagedVolumeCreationInfo, error) {
+	var output *template.ManagedVolumeCreationInfo
+	for name, volume := range input {
+		if output != nil {
```
Should we still return an error when `output != nil` if  `volume.EFS == nil` or `volume.EFS.UseManagedFS() == false` ?

bvtujo replied:
This should only be triggered now if there is more than one volume using Managed EFS in the list. Good catch.

Lou1415926 reviewed (commented):
Looks great! A few questions below ⬇️

efekarakus commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -251,41 +261,117 @@ func convertEFSPermissions(input map[string]manifest.Volume) ([]*template.EFSPer
+// See https://stackoverflow.com/a/14210379/5890422 for discussion of the possibility of collisions in CRC32 with
+// small numbers of hashes.
+func getRandomUIDGID(name string) uint32 {
+	return crc32.ChecksumIEEE([]byte(name))
```
should we do the suggestion here for `workloadtype/name`?

bvtujo replied:
I'm stripping the workloadtype logic out.

efekarakus commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -251,41 +261,117 @@ func convertEFSPermissions(input map[string]manifest.Volume) ([]*template.EFSPer
+			continue
+		}
+
+		if volume.EFS.UseManagedFS() {
```
nit:
```suggestion
		if !volume.EFS.UseManagedFS() {
			continue
		}
		uid := ...
```

efekarakus commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -298,11 +384,34 @@ func convertVolumes(input map[string]manifest.Volume) ([]*template.Volume, error
 		return nil, nil
 	}
+	// UID and GID should not be specified for non-managed volumes.
+	if !in.UseManagedFS() && !in.Config.IsEmpty() {
```
do we need to guard again against `in.UseManagedFS()`? Maybe we can make assumptions here saying the caller should ensure to call this function for non-managed FSs

efekarakus commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -251,41 +261,117 @@ func convertEFSPermissions(input map[string]manifest.Volume) ([]*template.EFSPer
+func getDirName(wl manifest.Workload) string {
+	for _, w := range manifest.ServiceTypes {
+		if aws.StringValue(wl.Type) == w {
+			return svcWlType + "/" + aws.StringValue(wl.Name)
```
Actually do we need the `workloadType/` prefix path? it looks like we store all these workloads under `components/` in SSM so it shouldn't be possible to declare a job with the same name as a service

bvtujo replied:
Huh, this is exposing an odd issue where we don't actually error if the name is taken. For example, in a workspace where I already have a LB Web Service called `fe`, I can do the following:

```
$ copilot job init -n fe -d ./Dockerfile --schedule "@every 90m"

✔ Manifest file for job fe already exists at copilot/fe/manifest.yml, skipping writing it.
Your manifest contains configurations like your container size and job schedule (@every 90m).

✔ Created ECR repositories for job fe.

Recommended follow-up actions:
- Update your manifest copilot/fe/manifest.yml to change the defaults.
- Run `copilot job deploy --name fe --env test` to deploy your job to a test environment.

$ copilot job deploy
✘ select job: no jobs found
```

I've captured this in #2156.

But it does look like we don't need the prefix after all!

bvtujo replied:
Removed the prefix.

efekarakus commented on internal/pkg/manifest/storage.go:
```diff
@@ -28,13 +38,87 @@ type SidecarMountPoint struct {
+
+// IsEmpty returns empty if the struct has all zero members.
+func (e *EFSVolumeConfiguration) IsEmpty() bool {
+	if e.FileSystemID == nil && e.RootDirectory == nil && e.AuthConfig == nil && e.UID == nil && e.GID == nil {
+		return true
+	}
+	return false
```
nit:
```suggestion
	return e.FileSystemID == nil && e.RootDirectory == nil && e.AuthConfig == nil && e.UID == nil && e.GID == nil 
```

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -251,41 +261,117 @@ func convertEFSPermissions(input map[string]manifest.Volume) ([]*template.EFSPer
+	if uid == nil && gid == nil {
+		return nil
+	}
+	if uid != nil && gid == nil {
+		return errInvalidUIDGIDConfig
+	}
+	if uid == nil && gid != nil {
+		return errInvalidUIDGIDConfig
+	}
```
Maybe 
```suggestion
	if (uid == nil) != (gid == nil) {
		return errInvalidUIDGIDConfig
	}
```

iamhopaul123 commented on internal/pkg/manifest/storage.go:
```diff
@@ -28,13 +38,87 @@ type SidecarMountPoint struct {
+	}
+	fromConfig := aws.StringValue(e.Config.FileSystemID)
+	if fromID != "" && fromID != fromConfig {
+		return nil, fmt.Errorf("read EFS ID: multiple values specified (%s, %s)", fromID, fromConfig)
```
Can we just ignore `e.ID` if `e.Config` is not empty? I remember we did the very similar thing with the other custom unmarshal.

iamhopaul123 commented on internal/pkg/manifest/storage.go:
```diff
@@ -28,13 +38,87 @@ type SidecarMountPoint struct {
 	AccessPointID *string `yaml:"access_point_id"` // Default ""
 }
+
+func contains(l []string, k string) bool {
```
We might need comment for this function or can we make the function name more specific reflecting it returns true regardless of capitalization?

bvtujo replied:
This function will go away in the next refactor PR; I vote to leave this as is if that's okay.

iamhopaul123 commented on internal/pkg/template/workload.go:
```diff
@@ -95,6 +95,10 @@ type StorageOpts struct {
 	ManagedVolumeInfo *ManagedVolumeCreationInfo // Used for delegating CreationInfo for Copilot-managed EFS.
 }
 
+func (s *StorageOpts) RequiresEFSCreation() bool {
```
Comment for public pls 🙏

iamhopaul123 commented on internal/pkg/template/workload.go:
```diff
@@ -118,9 +122,10 @@ type Volume struct {
+	GID     *uint32
 }
 
 type EFSVolumeConfiguration struct {
```
same here

iamhopaul123 commented on templates/workloads/partials/cf/volumes.yml:
```diff
@@ -1,11 +1,9 @@
-      FilesystemId:
-        Fn::ImportValue:
-          !Sub '${AppName}-${EnvName}-FilesystemID'
+      FilesystemId: !GetAtt EnvControllerAction.ManagedFileSystemID
```
I guess we haven't updated env controller for this. Is it safe for to use `!GetAtt EnvControllerAction.ManagedFileSystemID`?

bvtujo replied:
This is exported in the env stack and should be available in the Attributes of the EnvControllerAction. It's also gated behind a conditional on ManagedFileSystemInfo in the template so is only referenced when the Env Controller will have updated.

bvtujo committed:
chore: remove wlType from required input to convertStorageOpts

bvtujo committed:
chore: remove unused dirName function

bvtujo committed:
chore: address feedback

Add comment to avoid unnecessary safeguard for convertEFSConfig
Add comment for public function.

bvtujo committed:
chore: remove deadcode

bvtujo committed:
chore: address nits

bvtujo committed:
chore: add comment to EFSVolumeConfiguration

bvtujo committed:
chore: remove error from FSID

bvtujo committed:
chore: change quote style in error

Co-authored-by: Penghao He <penghaoh@amazon.com>

bvtujo committed:
Merge branch 'mainline' into managed-efs

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -40,6 +41,10 @@ var (
 	errNoSourceVolume  = errors.New(`"source_volume" cannot be empty`)
+
+	errUIDWithNonManagedFS = errors.New("UID and GID cannot be specified with non-managed EFS")
+	errInvalidUIDGIDConfig = errors.New("set managed filesystem access point creation info: must specify both UID and GID, or neither")
+	errReservedUID         = errors.New("set managed filesystem access point creation info: UID must not be 0")
```
Hmm why do we want to wrap this error? Maybe just
```suggestion
	errInvalidUIDGIDConfig = errors.New("must specify both UID and GID, or neither")
	errReservedUID         = errors.New("UID must not be 0")
```

bvtujo replied:
Will address in the refactor PR.

iamhopaul123 commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -251,41 +261,104 @@ func convertEFSPermissions(input map[string]manifest.Volume) ([]*template.EFSPer
+		}
+
+		if output != nil {
+			return nil, fmt.Errorf("validate managed EFS: cannot specify more than one managed volume per service")
```
Same with this error.

iamhopaul123 reviewed (approved):
LGTM overall with just two small nits. Feel free to move on.

mergify[bot] committed:
Merge branch 'mainline' into managed-efs
